### PR TITLE
RFC: add iframe login for js adapter

### PR DESCRIPTION
**Pull request by ibotty**
Hi!

To allow logging-in with iframes with the js-Adapter, I build that POC.  I rarely program javascript, so please bear with me.

It can be used as follows:

```javascript
var keycloak = Keycloak();
var iframe = document.getElementById('keycloak-login');
keycloak.init({ adapter: 'iframe', iframe: iframe  });
```

I am pretty sure, that I missed some things, but it apparently works pretty ok it my testing.

### CODE REVIEW

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 1]
```diff
@@ -1336,6 +1340,48 @@
...
+                iframe = initOptions.iframe;
+                return {
+                    login: function(options) {
+                        var promise = createPromise(false);
+
+                        console.log("login: ", options);
+                        console.log("iframe: ", iframe);
+                        iframe.addEventListener('load', function(event) {
+                            if (iframe.contentDocument) {
+                                var callback = parseCallback(iframe.contentDocument.location.href);
```
Comment (stianst) [inline adapters/oidc/js/src/main/resources/keycloak.js]: I'm guessing you are testing this with both Keycloak and the app on localhost? I believe iframe.contentDocument.location is only available if main window and iframe is on the same domain.

**Review by stianst — COMMENTED** [review]
Hi, thanks for your PR. I'm afraid there's quite a lot of work needed to be able to support embedding the login pages in the application like this. I'm also not convinced that it is a good thing to do. When embedded it is a lot harder for the user to see where the login page is coming from, and make sure they are only entering their credentials in Keycloak, and not directly in the application, which shouldn't have access to the credentials.

By default Keycloak does not allow embedding the login page in an iframe unless the application is hosted on the same origin, which in most cases it won't be. It would require a bigger change to how we set headers, and allowing embedding for specific clients.

The login theme today also doesn't really support this as it isn't designed to be embedded, but rather self-contained. Haven't tried this myself so don't know how bad it is.

We'd also need a way to test this automatically, as well as having documentation and quickstarts if we wanted to add this.

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 1 above)
Comment (ibotty) [reply adapters/oidc/js/src/main/resources/keycloak.js]: I am not testing on localhost, but I am testing with the web app and keycloak on the same domain (but a different subdomain).  I had to configure keycloak to set some security headers I don't recall though.

**Comment by ibotty** [comment]
Thank you for your comments.  I am aware that it is not as good as redirecting to another page.

Unfortunately the use case is a single-page app that has to be loaded all the time (think background music) and people logging in throughout interacting with the page.  I only see the two possibilities: loading keycloak in an iframe (as done here) and logging in directly within the app (and the app reimplementing parts of keycloak ui, speaking to keycloak).

Am I missing a flow that would suite that scenario?

**Comment by stianst** [comment]
Does your application require a user to be authenticated all the time? If so you'd just make sure people authenticate prior to starting using the app with the redirect.

**Comment by stianst** [comment]
It would probably be best to continue the discussion on the dev ML. Could you send a mail there?

**Comment by stianst** [comment]
Closing this if you are still interested in contributing around this, please open a thread on dev mailing list (see keycloak.org/community for more details).

**Pull request closed without merging**